dashboard/app/models/user.rb

```diff
@@ -1646,6 +1646,30 @@ def associate_with_potential_pd_enrollments
     end
   end
 
+  # When creating an account, we want to look for any channels that got created
+  # for this user before they signed in, and if any of them are in our Applab HOC
+  # course, we will create a UserScript entry so that they get a course card
+  def generate_progress_from_storage_id(storage_id, script_name='applab-intro')
+    # Find the set of levels this user started
+    # Worth noting that because ChannelToken uses levels (rather than script_levels)
+    # if a level is used in multiple scripts, we have no way to disambiguate which
+    # one a user saw it in, which becomes a challenge if we expand the scope of
+    # this beyond our applab-intro script.
+    channel_level_ids = ChannelToken.where(storage_id: storage_id).map(&:level_id)
+
+    levels_in_script = Script.get_from_cache(script_name).levels
+
+    # host_level will be self if we don't have a template level
+    # Expanding the scope beyond applab-intro would also be a challenge for template
+    # levels, as if a template is used in multiple scripts ,we have no way to know
+    # which a user saw it in
+    hoc_level_ids = levels_in_script.map(&:host_level).map(&:id)
```

Comment:
Note: I validated that our Script cache ensures that finding these level_ids results in no additional db queries.